bundles/org.openhab.voice.googlestt/src/main/java/org/openhab/voice/googlestt/internal/GoogleSTTService.java

```diff
@@ -0,0 +1,367 @@
+        this.oAuthFactory = oAuthFactory;
+        this.configAdmin = configAdmin;
+    }
+
+    @Activate
+    protected void activate(Map<String, Object> config) {
+        var serviceConfig = new org.openhab.core.config.core.Configuration(config).as(GoogleSTTConfiguration.class);
+        this.config = serviceConfig;
+        executor.submit(() -> GoogleSTTLocale.loadLocales(serviceConfig.refreshSupportedLocales));
+        String clientId = serviceConfig.clientId;
+        String clientSecret = serviceConfig.clientSecret;
+        if (!clientId.isEmpty() && !clientSecret.isEmpty()) {
+            var oAuthService = oAuthFactory.createOAuthClientService(SERVICE_PID, GCP_TOKEN_URI, GCP_AUTH_URI, clientId,
+                    clientSecret, GCP_SCOPE, false);
+            this.oAuthService = oAuthService;
+            if (!serviceConfig.oauthCode.isEmpty()) {
+                getAccessToken(oAuthService, serviceConfig.oauthCode);
+                deleteAuthCode();
+            }
+        }
+    }
+
+    @Override
+    public String getId() {
+        return SERVICE_ID;
+    }
+
+    @Override
+    public String getLabel(@Nullable Locale locale) {
+        return SERVICE_NAME;
+    }
+
+    @Override
+    public Set<Locale> getSupportedLocales() {
+        return GoogleSTTLocale.getSupportedLocales();
+    }
+
+    @Override
+    public Set<AudioFormat> getSupportedFormats() {
+        return Set.of(
+                new AudioFormat(AudioFormat.CONTAINER_WAVE, AudioFormat.CODEC_PCM_SIGNED, false, 16, null, 16000L),
+                new AudioFormat(AudioFormat.CONTAINER_OGG, "OPUS", null, null, null, 8000L),
+                new AudioFormat(AudioFormat.CONTAINER_OGG, "OPUS", null, null, null, 12000L),
+                new AudioFormat(AudioFormat.CONTAINER_OGG, "OPUS", null, null, null, 16000L),
+                new AudioFormat(AudioFormat.CONTAINER_OGG, "OPUS", null, null, null, 24000L),
+                new AudioFormat(AudioFormat.CONTAINER_OGG, "OPUS", null, null, null, 48000L));
+    }
+
+    @Override
+    public STTServiceHandle recognize(STTListener sttListener, AudioStream audioStream, Locale locale,
+            Set<String> set) {
+        var scheduledTask = backgroundRecognize(sttListener, audioStream, locale, set);
+        return () -> scheduledTask.cancel(true);
+    }
+
+    private void getAccessToken(OAuthClientService oAuthService, String oauthCode) {
+        logger.debug("Trying to get access and refresh tokens.");
+        try {
+            oAuthService.getAccessTokenResponseByAuthorizationCode(oauthCode, GCP_REDIRECT_URI);
+        } catch (OAuthException | OAuthResponseException e) {
+            logger.debug("Error fetching access token: {}", e.getMessage(), e);
+            logger.error("Error fetching access token. Invalid oauth code? Please generate a new one.");
+        } catch (IOException e) {
+            logger.error("An unexpected IOException occurred: {}", e.getMessage());
```

Comment:
WARN level is enough. Please avoid ERROR level in all the class.

Rather than 2 consecutive calls to logger, you could use an if then else with logger.isDebugEnabled
In case of IOException, please inject a context in the message, something like: An unexpected IOException occurred when fetching access token: {}